src/callbacks/sanitycheck.jl

```diff
@@ -0,0 +1,150 @@
+
+struct Check
+    checkfn
+    name
+    error::Bool
+    message
+end
+
+
+struct SanityCheckException <: Exception end
+
+Base.show(io::IO, check::Check) = print(io, "Check(\"", check.name, "\")")
+
+
+function runchecks(checks, learner)
+    failed = trues(length(checks))
+    for (i, check) in enumerate(checks)
+        failed[i] = !check.checkfn(learner)
+    end
+
+    failedchecks = checks[failed]
+
+    if isempty(failedchecks)
+        return
+    end
+
+    println("$(length(failedchecks))/$(length(checks)) sanity checks failed:")
+
+    for (i, check) in enumerate(failedchecks)
+        println("---")
+        println(i, ": ", check.name, " (", check.error ? "ERROR" : "WARNING", ")")
+        println()
+        println(check.message)
+    end
+
+    if any(getfield.(failedchecks, :error))
+        throw(SanityCheckException())
+    end
+end
+
+
+"""
+    SanityCheck([checks; usedefault = true])
+
+Callback that runs sanity [`Check`](#)s when the `Learner` is initialized.
+If `usedefault` is `true`, it will run all checks in FluxTraining.CHECKS
+in addition to the ones you pass in.
+"""
+struct SanityCheck <: Callback
+    checks::Vector{Check}
+    function SanityCheck(checks = []; usedefault = true)
+        if usedefault
+            checks = vcat(checks, CHECKS)
+        end
+        return new(checks)
+    end
+end
+
+
+
+stateaccess(::SanityCheck) = (
+    data = Read(),
+    model = Read(),
+    lossfn = Read(),
+    optimizer = Read(),
+    callbacks = Read(),
+)
+
+
+function on(::Init, phase::AbstractTrainingPhase, cb::SanityCheck, learner)
+    runchecks(cb.checks, learner)
+end
+
+
+# Checks
+
+
+const CHECKS = [
+    Check(
+        "Has training data iterator",
+        true,
+"""
+`learner` does not have a training data iterator. You can pass it to
+`Learner` in one of the following ways:
+
+- `Learner(model, traindataiter, opt, lossfn)`
+- `Learner(model, (traindataiter, valdataiter), opt, lossfn)`
+- `Learner(model, (training = traindataiter,), opt, lossfn)`
+"""
+        ) do learner
+        !isnothing(learner.data.training)
+    end,
+    Check(
+        "Has validation data iterator",
+        false,
+"""
+`learner` does not have a validation data iterator. This means
+you won't be able to fit `ValidationPhase`s.
+
+You can pass it to `Learner` like this:
+
+- `Learner(model, (traindataiter, valdataiter), opt, lossfn)`
+
+Or if you want to use training data as validation data:
+
+- `Learner(model, (traindataiter, traindataiter), opt, lossfn)`
+"""
+        ) do learner
+        !isnothing(learner.data.validation)
+    end,
+    Check(
+        "Data iterators iterate over tuples",
+        true,
+"""
+Data iterators need to be iterable and return tuples.
+This means that `for (x, y) in dataiter end` works where
+`(x, y)` is a pair of model inputs and outputs.
+"""
+        ) do learner
+        try
+            batch, _ = Base.iterate(learner.data.training)
```

Comment:
Is there something that would allow this to handle stateful iterators like those exposed in MLDataPattern? i.e. not consume the first batch outright.

Reply:
I think this isn't a problem since they can be iterated over again.

Calling `batch, state = iterate(iter)` will iterate from the start while `batch', state' = iterate(iter, state)` would continue iteration from the previous state.

This does mean that it won't work for iterables that can only be iterated through once, like generators. But since one usually wants to train for multiple epochs, the data iterators have to support repeated iteration anyways.

Reply:
`Iterators.Stateful(...)` doesn't, but to your point I imagine most are going to be using something that is resumable if they want to train for more than one epoch!